File: 496-NextGreaterElementI/496-NextGreaterElementI.java
// Last updated: 09/07/2026, 10:07:51
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap<Integer,Integer> map = new HashMap<>();
        int n = nums2.length;
        Stack<Integer> stack = new Stack<>();
        for(int i = n-1;i>=0;i--){
            while(!stack.isEmpty() && nums2[i] >= stack.peek()){
                stack.pop();
            }
            if(stack.isEmpty()) map.put(nums2[i],-1);
            else{
                map.put(nums2[i],stack.peek());
            }
            stack.push(nums2[i]);
        }

        int n1 = nums1.length;
        int result[] = new int[n1];
        for(int i = 0;i<n1;i++){
            result[i] = map.get(nums1[i]);
        }
        return result;
    }
}

File: 496-NextGreaterElementI/496-NextGreaterElementI.py
# Last updated: 09/07/2026, 10:07:56
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        hm = {}
        for i in range(len(nums2)-1,-1,-1): 
            while stack and stack[-1] <= nums2[i]:
                stack.pop()
            
            if len(stack) == 0:
                hm[nums2[i]] = -1
            else:
                hm[nums2[i]] = stack[-1]
            stack.append(nums2[i])        
        ans = []
        for j in range(len(nums1)):
            ans.append(hm[nums1[j]])     
        return ans    

